### ISSUE TITLE
Best way to partition an API

### ISSUE THREAD

**Issue by gmsDave** [domaindrivendev/Ahoy#198]
We are starting to develop what will be a large API and are already experiencing conflicting schemaIds.  I would like to subdivide my API by business domain context.  We started by using Namespaces to isolate my API controllers and models from each other, but they are still contained within a single ASP.NET Core MVC application and as a result, swashbuckle attempts to generate a single swagger.json but throws an exception about Conflicting schemaIds.  I am not using the UseFullTypeNameInSchemaIds config property (which is missing in 6.0.0-beta902) or the CustomSchemaIds function, and I don't think I want to.  

What I would prefer is a way to define separate swagger.json files for each namespace, and then providing a table of contents, of sorts, for each API context.  In this way, we have the option to version each API by context, rather than a single version increment en masse when possibly only one API resource was changed.

Is this possible with swashbuckle or would I need to have separate ASP.NET Core MVC project for each isolated namespace?  My issue will that approach, aside from the duplication in project settings and configuration, plus the added complexity to continuous integration and deployment, is the developer experience with each project running on different port on their dev boxes.

**Comment by rynowak** [comment]
MVC supports setting an arbitrary `GroupName` for each API that's *document* with `ApiDescription` https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ApiExplorerModel.cs

I think by default we make each controller it's own group, but the intent is that you can write a convention to set the group name programmatically. Example [here](https://github.com/aspnet/Mvc/blob/760c8f38678118734399c58c2dac981ea6e47046/test/WebSites/ApiExplorerWebSite/ApiExplorerVisibilityEnabledConvention.cs). What you'd probably want to do is write a convention that programmatically maps your controller namespace to a group name.

As far as configuring swashbuckle to fork on group name, @domaindrivendev will have to help with that, but this is definitely something that group name was intended for.

**Comment by SimonTouchtech** [comment]
If I understand you correctly you want to create multiple swagger.json files depending on something in your code (namespace in this case). This can be done in 6.0.0-beta902 by using the `MultipleApiVersions` method on `SwaggerGenOptions`. The second argument of that method is a `Func` taking an `ApiDescription` and a `string` that should return `true` if the `ApiDescription` describes an API that's part of the "version" indicated by the string.

As for the table of contents, my guess is you would have to create that yourself, but perhaps @domaindrivendev knows some other way.

**Note:** This has been slightly changed in the latest code (that's not available on Nuget) with the support for multiple versions instead being considered support for multiple documents, which is more semantically correct with what you want. See https://github.com/domaindrivendev/Ahoy/commit/391127571fd0bb659bd31e1f9f262df4de263465

**Comment by domaindrivendev** [comment]
@SimonTouchtech is spot on. To see this in action, you can check out the "MultipleVersions" example Website:

https://github.com/domaindrivendev/Ahoy/blob/master/test/WebSites/MultipleVersions/Startup.cs

This partitions by version but because it's based on a lambda which you provide, you can break down your documentation however you like. You will need to specify which Swagger documents you'd like to use in the swagger-ui:

https://github.com/domaindrivendev/Ahoy/blob/master/test/WebSites/MultipleVersions/Startup.cs#L62

When you provide multiple documents (as in the example) the UI will automatically display a drop-down in the top right to switch between documents. This would essentially be your table of contents.

**Comment by gmsDave** [comment]
Thanks @SimonTouchtech and @domaindrivendev.  I have been sidetracked lately getting OAuth 2.0 working across all layers of our new app with Identity Server 4, Angular 2 and WebAPI, I have had to backburner this requirement for now.  We're pushing for a release in December and are under the gun.  We moved this down in the backlog as we aren't publishing the API for external customers until next year and we are working around it for now using unique class names across the namespaces.  Not ideal, but it's working for us for now.
